File: sdks/cpp/examples/lobby_and_challenge.cpp
#include "../fedichess_client.hpp"
#include <chrono>
#include <iostream>
#include <string>
#include <cstdlib>

int main() {
  std::string bridge_dir = FEDICHESS_BRIDGE_DIR;
  std::string bridge_path = bridge_dir + "/dist/index.js";
  std::string bridge_cwd = bridge_dir;

  const char* env_bridge = std::getenv("FEDICHESS_BRIDGE");
  if (env_bridge) bridge_path = env_bridge;
  else if (bridge_dir.empty()) bridge_path = "node";

  fedichess::FediChessClient client(bridge_path, bridge_cwd);
  if (!client.start_bridge()) {
    std::cerr << "Failed to start bridge\n";
    return 1;
  }

  auto r = client.join_lobby();
  if (!r.value("ok", false)) {
    std::cerr << "join_lobby failed: " << r.dump() << "\n";
    return 1;
  }

  auto ts = std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::system_clock::now().time_since_epoch())
                .count();
  client.send("heartbeat",
              {{"id", "sdk-cpp-example"},
               {"elo", 1200},
               {"name", "CppSDK"},
               {"ready", true},
               {"timestamp", ts}});

  for (int i = 0; i < 3; ++i) {
    auto ev = client.poll_event(2.0);
    if (!ev.is_null())
      std::cout << "Event: " << ev.value("event", "") << " "
                << ev.value("peerId", "") << " " << ev.dump() << "\n";
  }

  auto peers = client.get_peers();
  std::cout << "Peers in lobby: ";
  for (const auto& p : peers) std::cout << p << " ";
  std::cout << "\n";

  if (!peers.empty()) {
    std::string game_id = "00000000-0000-0000-0000-000000000001";
    ts = std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::system_clock::now().time_since_epoch())
             .count();
    client.send("challenge",
                {{"type", "challenge"},
                 {"gameId", game_id},
                 {"challengerId", "sdk-cpp-example"},
                 {"challengerName", "CppSDK"},
                 {"challengerElo", 1200},
                 {"color", "w"},
                 {"timestamp", ts}},
                peers[0]);
    std::cout << "Sent challenge to " << peers[0] << " gameId " << game_id << "\n";
    for (int i = 0; i < 10; ++i) {
      auto ev = client.poll_event(2.0);
      if (!ev.is_null() && ev.value("event", "") == "challResp") {
        std::cout << "challResp: " << ev.dump() << "\n";
        break;
      }
    }
  }

  client.leave_lobby();
  client.stop();
  std::cout << "Done.\n";
  return 0;
}

File: sdks/python/examples/lobby_and_challenge.py
#!/usr/bin/env python3
"""
Example: join lobby, send heartbeat, list peers, optionally send a challenge.
Run the bridge first (see sdks/bridge/README.md), then from repo root or sdks/python:
  python -m fedichess.examples.lobby_and_challenge
  or: python sdks/python/examples/lobby_and_challenge.py
"""
import os
import sys
import time

# Add parent so we can import fedichess
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from fedichess import FediChessClient


def main() -> None:
    # Prefer bridge next to this repo (sdks/bridge/dist/index.js)
    # From sdks/python/examples/ -> sdks/bridge
    bridge_dir = os.path.join(os.path.dirname(__file__), "..", "..", "..", "bridge")
    bridge_path = os.path.join(bridge_dir, "dist", "index.js")
    if not os.path.isfile(bridge_path):
        bridge_path = os.environ.get("FEDICHESS_BRIDGE", "node")

    client = FediChessClient(bridge_path=bridge_path, bridge_cwd=bridge_dir if os.path.isdir(bridge_dir) else None)
    client.start_bridge()

    r = client.join_lobby()
    if not r.get("ok"):
        print("join_lobby failed:", r.get("error", r))
        return

    # Send heartbeat so we appear in the lobby
    client.send("heartbeat", {
        "id": "sdk-python-example",
        "elo": 1200,
        "name": "PythonSDK",
        "ready": True,
        "timestamp": int(time.time() * 1000),
    })

    # Drain a few events (optional)
    for _ in range(3):
        ev = client.poll_event(timeout=2.0)
        if ev:
            print("Event:", ev.get("event"), ev.get("peerId", ""), ev.get("payload"))

    peers = client.get_peers()
    print("Peers in lobby:", peers)

    if peers:
        # Optional: send a challenge to the first peer
        import uuid
        game_id = str(uuid.uuid4())
        client.send("challenge", {
            "type": "challenge",
            "gameId": game_id,
            "challengerId": "sdk-python-example",
            "challengerName": "PythonSDK",
            "challengerElo": 1200,
            "color": "w",
            "timestamp": int(time.time() * 1000),
        }, peer_id=peers[0])
        print("Sent challenge to", peers[0], "gameId", game_id)
        # Wait for challResp
        for _ in range(10):
            ev = client.poll_event(timeout=2.0)
            if ev and ev.get("event") == "challResp":
                print("challResp:", ev.get("payload"))
                break

    client.leave_lobby()
    client.stop()
    print("Done.")


if __name__ == "__main__":
    main()
